clientv3/balancer.go

```diff
@@ -354,3 +358,55 @@ func getHost(ep string) string {
 	}
 	return url.Host
 }
+
+func (b *simpleBalancer) switchEndpoint() {
+	allEps := []string{}
+	allEpsWithoutPinAddr := []string{}
+
+	b.mu.RLock()
+	prevPinAddr := b.pinAddr
+	for addr, ep := range b.host2ep {
+		allEps = append(allEps, ep)
+		if addr != b.pinAddr {
+			allEpsWithoutPinAddr = append(allEpsWithoutPinAddr, ep)
+		}
+	}
+	lastSwitchTime := b.lastSwitchTime
+	b.mu.RUnlock()
+
+	if len(allEpsWithoutPinAddr) == 0 {
+		// only one available endpoint, do not switch
+		return
+	}
+
+	tNow := time.Now()
+	if lastSwitchTime.Add(30 * time.Second).After(tNow) {
+		// avoid too frequently switch
+		return
+	} else {
+		b.mu.Lock()
+		b.lastSwitchTime = tNow
+		b.mu.Unlock()
+	}
+
+	// may face some race condition here
+	// and also at update allEps back
+	// but be optimistic, auto sync can refresh the right eps periodically
+	b.updateAddrs(allEpsWithoutPinAddr)
+
+
+	// at most sleep 10 seconds here
```

Comment:
depending on sleeps is slow and unreliable; the balancer probably needs to directly reason about graylisted endpoints